doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+# Reference Counting
+
+## Problem
+
+Elektra's data structures are shared between multiple owners.
+A `Key` can be owned by many different `KeySet`s.
+Both `Key` and `KeySet` can be used in bindings.
+If the languages of those bindings have automatic memory management, this also implies a shared ownership.
+
+To solve this issue of shared ownership, we need reference counting.
+
+An adjacent problem is that the names of `Key`s must be read-only while the `Key` is part of a `KeySet`.
+This can also be solved by a secondary counter just for `KeySet`s.
+
+There has also been [discussion](copy_on_write.md) about adding further data structures, which would need reference counters as well.
+
+## Constraints
+
+- Deleting data structures that still have references must be prevented
+- The APIs for `ksAppendKey` and `ksNew` must remain ergonomic.
+  The reference counting must not prevent the inline creation of a `Key` while adding it to a `KeySet`.
```

Comment:
Better move this to assumptions for now (as long as we don't have use cases).

Reply:
What do you mean with "as long as we don't have use cases"?

```c
keyAppendKey (ks, keyNew ("/foo", KEY_END));
ksNew (1, keyNew ("/foo", KEY_END), KS_END);
```

That's what I'm talking about. A very real use case and integral part of the current API. The API might change, so maybe we remove the specific references to the functions, but I think we all agree that we want to keep this option. The alternative would be something far more verbose like (placeholder API):

```c
Key * k  = keyNew ("/foo", KEY_END);
keyAppendKey (ks, k);
keyDel (k);

Key * k1  = keyNew ("/foo", KEY_END);
ksNew (1, k1, KS_END);
keyDel (k);
```

I also don't know how this could be phrased as an assumption.

Reply:
The question is if we want to support the use case to be able to construct any keyset with a single function call. (Looking at our API we probably want, but we don't have documented use cases for that.)

Reply:
Still, either we leave it as a constraint or remove it completely. It doesn't make sense as an assumption. IMO assumptions are statements that simplify the problem space. This is exactly the opposite. It would be stupid of us to assume something that makes our lives more difficult.

Reply:
Ok, I agree, let us keep it as constraint. We cannot write use cases for every detail.